### PR TITLE
Add Canonicalized_Extension_Headers parameter to storage signed URLs.

### PR DESCRIPTION
**Pull request by calavera**
Hi!

I was about to use the cloud storage signed URLs when I realized
that this library doesn't implement extension headers.

This change allows signed URLs issuers to send extension headers.
The implementation follows the spec in the documentation:

https://cloud.google.com/storage/docs/access-control/signed-urls#about-canonical-extension-headers

However, it assumes that the issuer will normalize duplicated headers
keys into a unique header with values separated by comma.

I'd really appreciate your feedback.

Signed-off-by: David Calavera <david.calavera@gmail.com>

### CODE REVIEW

**Comment by blowmage** [comment]
Thanks for the PR @calavera! We will get this reviewed ASAP!

**Comment by blowmage** [comment]
Hey @calavera, thanks again for this PR. I found some issues and submitted a PR for your PR. Pull request inception, we must go deeper.

deeper

The needed changes are pretty straightforward. The headers need to be sorted and downcased. The headers are to be used in the signature hash and not added to the URL query parameters.

Let me know if you have any questions. I've played around with the acceptance test and I'm reasonably confident the fix in my PR works as expected.

**Comment by calavera** [comment]
Thanks a lot for your help @blowmage. I forgot about those required changes :raised_hands: 

I'm not sure why the bot is giving that error though.

**Comment by blowmage** [comment]
I am the author of the commit.

Googlebot is a little paranoid at time. 😉

**Comment by blowmage** [comment]
@quartzmo Can you look this over and approve/reject. I approved but I also authored one of the commits.

File: google-cloud-storage/lib/google/cloud/storage/bucket.rb [hunk 1]
```diff
@@ -665,11 +665,23 @@ def create_file file, path = nil, acl: nil, cache_control: nil,
...
+        # @example Using the headers option:
+        #   require "google/cloud/storage"
+        #
+        #   storage = Google::Cloud.storage
+        #
+        #   bucket = storage.bucket "my-todo-app"
+        #   shared_url = bucket.signed_url "avatars/heidi/400x400.png",
+        #                                  headers: {
+        #                                    "x-goog-acl" => "private",
+        #                                    "x-goog-meta-foo" => "bar,baz"
```
Comment (quartzmo) [inline google-cloud-storage/lib/google/cloud/storage/bucket.rb]: @blowmage You changed the headers in the acceptance test to be capitalized (`X-Goog-ACL`), do you want to do it as well in these docs examples?

Comment (blowmage) [inline google-cloud-storage/lib/google/cloud/storage/bucket.rb]: The headers will be downcased, per the [spec](https://cloud.google.com/storage/docs/access-control/signed-urls#about-canonical-extension-headers). That's why the various tests are using different casing, to ensure that it works correctly.

I don't know how the headers should be presented in the documentation, however. What do you think?

Comment (quartzmo) [inline google-cloud-storage/lib/google/cloud/storage/bucket.rb]: Correct, the headers should appear in lower case like the cloud.google.com docs. Thanks! Merging.

**Comment by blowmage** [comment]
🎉

**Comment by blowmage** [comment]
@calavera Thanks for this PR! We will release this as soon as we get passing builds across Linux, OS X, and Windows.

**Comment by blowmage** [comment]
@calavera This had been released in 0.23.1. Thanks again!

**Comment by calavera** [comment]
awesome, thanks! :confetti_ball: :tada:

**Pull request merged**